STM32 Flash fix and optimizations

erwango opened the pull request:
Fix in STM32l4 driver that could be vulnerable to unaligned memory access if data is not aligned in flash_write operation.

Additionally, rework flash registers definitions for STM32F1/STM32F3 (missing volatile instruction).

erwango committed:
drivers/flash: stm32f3/stm32f1: define flash registers as volatile

On STM32F3 devices, a trick was required for normal use of registers.
This was actually an issue in flash registers defintions which should
be defined as volatile.
Fix this and additionaly, fix definition for STM32F1 which was also
lacking the volatile instruction.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -149,9 +149,15 @@ int flash_stm32_write_range(struct device *dev, unsigned int offset,
-		rc = write_dword(dev, offset, ((const u64_t *) data)[i>>3]);
+		/* Use local variable to access 'data' word by word */
+		/* to avoid errors in case 'data' is unaligned */
+		dword = ((u64_t)(tmp + (i >> 2))[1] << 32) +
```
Would UNALIGNED_GET be useful here?

erwango replied:
Let me have a test

erwango replied:
Seems like UNALIGNED_GET doesn't like the expression it should be applied to.

aurel32 replied:
`UNALIGNED_GET` needs a pointer in argument, so you need to use UNALIGNED_GET((const u64_t *) data + (i >> 3))``

erwango replied:
Right! fixed,  thanks again.

aurel32 commented on drivers/flash/flash_stm32.c:
```diff
@@ -210,11 +210,9 @@ static int flash_stm32_write_protection(struct device *dev, bool enable)
+			/* Order of operations is critical here */
+			/* Add a barrier to avoid compiler optimisations */
+			__ISB();
 			regs->keyr = FLASH_KEY2;
```
I am not sure that you need an instruction barrier, the flash registers are in a strongly ordered region, so the CPU should already ensure that the two writes actually reach the registers.

OTOH you might want a compiler barrier, but that should not be needed if the registers are declared `volatile`. And I just realized that all the flash registers are declared as volatile, except on the STM32F1x and STM32F3x. Could it be the real issue?

erwango replied:
>  Could it be the real issue?

This sounds more realistic indeed. I'll have a try.

erwango replied:
That was it, thanks for the hint!

aurel32 commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -149,9 +149,15 @@ int flash_stm32_write_range(struct device *dev, unsigned int offset,
+		dword = ((u64_t)(tmp + (i >> 2))[1] << 32) +
+			 (u64_t)(tmp + (i >> 2))[0];
+		rc = write_dword(dev, offset, dword);
 		if (rc < 0) {
```
That part looks all good.

erwango committed:
drivers/flash: stm32l4: Prepare for unaligned accesses in flash writes

On STM32L4, flash writes operations are performed 8 bytes at a time.
Though, it is possible that *data in flash_write functions is not
aligned. To avoid issues, use UNALIGNED_GET macro to access *data.

Issue has been detected using settings subsystem on STM32WB, which has
same 8bytes write block size. The patch is extended to STM32L4 series
for same reason.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

aurel32 reviewed (approved):
Thanks for the update, LGTM.
